support/k8s/spire-adm-webhook/webhook.go

```diff
@@ -0,0 +1,225 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	admv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type WebhookConfig struct {
+	Log         logrus.FieldLogger
+	R           registration.RegistrationClient
+	TrustDomain string
+	Cluster     string
+	PodLabel    string
+}
+
+type Webhook struct {
+	c WebhookConfig
+}
+
+func NewWebhook(config WebhookConfig) *Webhook {
+	return &Webhook{
+		c: config,
+	}
+}
+
+func (w *Webhook) Initialize(ctx context.Context) error {
+	// ensure there is a node registration entry for PSAT nodes in the cluster.
+	return w.createEntry(ctx, &common.RegistrationEntry{
+		ParentId: idutil.ServerID(w.c.TrustDomain),
+		SpiffeId: w.nodeID(),
+		Selectors: []*common.Selector{
+			{Type: "k8s_psat", Value: fmt.Sprintf("cluster:%s", w.c.Cluster)},
+		},
+	})
+}
+
+func (w *Webhook) ReviewAdmission(ctx context.Context, req *admv1beta1.AdmissionRequest) (*admv1beta1.AdmissionResponse, error) {
+	w.c.Log.WithFields(logrus.Fields{
+		"namespace": req.Namespace,
+		"name":      req.Name,
+		"kind":      req.Kind.Kind,
+		"version":   req.Kind.Version,
+		"operation": req.Operation,
+	}).Debug("ReviewAdmission called")
+
+	if err := w.reviewAdmission(ctx, req); err != nil {
+		return nil, err
+	}
+
+	return &admv1beta1.AdmissionResponse{
+		UID:     req.UID,
+		Allowed: true,
+	}, nil
+}
+
+// reviewAdmission handles CREATE and DELETE requests for pods in
+// non-kubernetes namespaces. Ideally the ValidatingAdmissionWebhook
+// configuration has filters in place that prevent the webhook from being
+// called for requests other than these but just in case, we filter here as
+// well.
+func (w *Webhook) reviewAdmission(ctx context.Context, req *admv1beta1.AdmissionRequest) error {
+	switch req.Namespace {
+	case metav1.NamespacePublic, metav1.NamespaceSystem:
+		return nil
+	}
+
+	if req.Kind != (metav1.GroupVersionKind{Version: "v1", Kind: "Pod"}) {
+		return nil
+	}
+
+	switch req.Operation {
+	case admv1beta1.Create:
+		pod := new(corev1.Pod)
+		if err := json.Unmarshal(req.Object.Raw, pod); err != nil {
+			return errs.New("unable to unmarshal %s/%s object: %v", req.Kind.Version, req.Kind.Kind, err)
+		}
+		return w.createPodEntry(ctx, pod)
+	case admv1beta1.Delete:
+		return w.deletePodEntry(ctx, req.Namespace, req.Name)
+	}
+
+	return nil
+}
+
+func (w *Webhook) createPodEntry(ctx context.Context, pod *corev1.Pod) error {
+	if w.c.PodLabel != "" {
+		// the webhook has been been configured with a pod label. if the pod
+		// has that label, use the value to construct the pod entry. otherwise
+		// ignore the pod altogether.
+		if labelValue, ok := pod.Labels[w.c.PodLabel]; ok {
+			return w.createPodEntryByLabel(ctx, pod, w.c.PodLabel, labelValue)
+		}
+		return nil
+	}
+
+	// the webhook has not been configured with a pod label. create an entry
+	// based on the service account.
+	return w.createPodEntryByServiceAccount(ctx, pod)
```

Comment:
At least in my environment, I'd like to be able to disable this behavior -- Could we put this behind a flag?

Reply:
I guess it kind of already is, in that if you configure a pod label it won't create the entry based on service account?

Reply:
I want to be able to have pods with no registered spiffe identity, instead of a bogus one that doesn't match our naming scheme.

Reply:
Ignore my comment!  I realized I misread the code.  It acts as I desire :).